salt/_modules/caasp_cri.py

```diff
@@ -0,0 +1,389 @@
+
+    name
+        Name of the container. This is checked against the ``metadata.name``
+        field of a kubernetes pod.
+
+    namespace
+        Name of the namespace to search the container inside.
+
+    command
+        An array defining the command to run
+
+    .. code-block:: bash
+
+        salt '*' caasp_cri.exec_cmd_inside_of_container container_id='boring-elon' command='bash -c "cat /etc/hosts-caasp > /etc/hosts"'
+
+    '''
+    container_id = get_container_id(name, namespace)
+
+    if container_id is None:
+        return {'success': False,
+                'info': 'Cannot find specified container'}
+
+    cmd = "crictl --runtime-endpoint {socket} exec -s {container_id} {command}".format(
+            socket=cri_runtime_endpoint(),
+            container_id=container_id,
+            command=command)
+    result = __salt__['cmd.run_all'](cmd,
+                                     output_loglevel='trace',
+                                     python_shell=False)
+
+    info = ''
+    success = False
+
+    if result['retcode'] != 0:
+        info = result['stderr']
+    else:
+        info = result['stdout']
+        success = True
+
+    return {'success': success, 'info': info}
+
+
+def cri_service_name():
+    '''
+    Return a string holding the name of the service identifying the CRI.
+
+    This is used internally by our salt states to DRY them.
+    '''
+    if 'admin' in __salt__['grains.get']('roles', []):
+        return 'docker'
+
+    cri = __salt__['pillar.get']('cri:name', 'docker').lower()
+
+    if cri == 'docker':
+        return 'docker'
+    elif cri == 'crio':
+        return 'crio'
+    else:
+        raise InvalidConfigError(
+                'Uknown CRI specified inside of pillars: {}'.format(cri))
+
+
+def cri_salt_state_name():
+    '''
+    Return a string holding the name of the salt state that manages the CRI.
+
+    This is used internally by our salt states to DRY them.
+    '''
+    return cri_service_name()
+
+
+def cri_package_name():
+    '''
+    Return a string holding the name of the package providing the CRI
+
+    This is used internally by our salt states to DRY them.
+    '''
+    if 'admin' in __salt__['grains.get']('roles', []):
+        return 'docker'
+
+    cri = __salt__['pillar.get']('cri:name', 'docker').lower()
+
+    if cri == 'docker':
+        return 'docker'
+    elif cri == 'crio':
+        return 'cri-o'
+    else:
+        raise InvalidConfigError(
+                'Uknown CRI specified inside of pillars: {}'.format(cri))
+
+
+def cri_runtime_endpoint():
+    '''
+    Return the path to the socket required by crictl to communicate
+    with the CRI
+    '''
+    if 'admin' in __salt__['grains.get']('roles', []):
+        return '/var/run/dockershim.sock'
+
+    cri = __salt__['pillar.get']('cri:name', 'docker').lower()
+
+    if cri == 'docker':
+        return '/var/run/dockershim.sock'
+    elif cri == 'crio':
+        return '/var/run/crio/crio.sock'
```

Comment:
Nitpick: This looks like something we could configure with grains? We reference the endpoints in some config files as well.

Reply:
Grains are about collecting facts from the nodes (eg: arch, cpu,...). In this case I think it's fine to have the code structured in that way. But maybe I misunderstood your comment...

Reply:
Oups, sorry. I am not familiar with Salt and interpreted grains to be some kind of local key-value store.